src/ConfCrypt/Template.hs

```diff
@@ -0,0 +1,70 @@
+module ConfCrypt.Template (
+    renderTemplate
+    ) where
+
+import Text.Megaparsec
+import Text.Megaparsec.Char
+import Data.Text
+import ConfCrypt.Types (Parameter(..))
+import qualified Data.Text as T
+
+-- "text %k=%v %%"
+-- "text foo=bar %"
+-- ["text ", k, "=", v, " %"]
+-- %[a-z] %x %y %z
+
+
+renderTemplate :: Text -> Parameter -> Text
+renderTemplate template param = case parse parseTemplate "" template of
```

Comment:
This may be parsing the `template` on every single call. To prevent this, you can first parse the template, and then return a closure that closes over the result. Something like this:

```
renderTemplate template = case parse parseTemplate "" template of
    Left err -> const . pack . show $ err
    Right parsed -> \param -> foldMap (replaceVars param) parsed
    where
        replaceVars _ (Text_ t) = t
        replaceVars param (Variable_ Name) = paramName param
        ...
```

This way, the parsing will happen when you call `renderTemplate tpl` (at use site), and not for every value.

It is very possible that GHC would make this optimization automatically, and also, in this particular context it doesn't really matter that much. So this more of an educational note.

Reply:
If you were writing this code would you write it with an explicit closure, or would you wait and see if it was a problem first?

Reply:
In this situation, it will never become a problem, because the volumes are tiny. But I think I would still do an explicit closure, just because I'm OCD like that.